model/map/src/main/java/org/keycloak/models/map/realm/MapRealmProvider.java

```diff
@@ -0,0 +1,469 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.jboss.logging.Logger;
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientScopeModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RealmModel.SearchableFields;
+import org.keycloak.models.RealmProvider;
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.map.common.Serialization;
+import org.keycloak.models.map.common.StreamUtils;
+import org.keycloak.models.map.common.StreamUtils.Pair;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder.Operator;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+public class MapRealmProvider implements RealmProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapRealmProvider.class);
+    private final KeycloakSession session;
+    final MapKeycloakTransaction<String, MapRealmEntity, RealmModel> tx;
+    private final MapStorage<String, MapRealmEntity, RealmModel> realmStore;
+
+    public MapRealmProvider(KeycloakSession session, MapStorage<String, MapRealmEntity, RealmModel> realmStore) {
+        this.session = session;
+        this.realmStore = realmStore;
+        this.tx = new MapKeycloakTransaction<>(realmStore);
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private Function<MapRealmEntity, RealmModel> entityToAdapterFunc() {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+
+        return origEntity -> new MapRealmAdapter(session, registerEntityForChanges(origEntity));
+    }
+
+    private MapRealmEntity registerEntityForChanges(MapRealmEntity origEntity) {
+        final MapRealmEntity res = Serialization.from(origEntity);
+        tx.updateIfChanged(origEntity.getId(), res, MapRealmEntity::isUpdated);
+        return res;
+    }
+
+    @Override
+    public RealmModel createRealm(String name) {
+        return createRealm(KeycloakModelUtils.generateId(), name);
+    }
+
+    @Override
+    public RealmModel createRealm(String id, String name) {
+        if (getRealmByName(name) != null) {
+            throw new ModelDuplicateException("Realm with given name exists: " + name);
+        }
+
+        final String entityId = id == null ? KeycloakModelUtils.generateId() : id;
+
+        LOG.tracef("createRealm(%s, %s)%s", id, name, getShortStackTrace());
+
+        MapRealmEntity entity = new MapRealmEntity(entityId);
+        entity.setName(name);
+        if (tx.read(entity.getId()) != null) {
```

Comment:
This can be done before MapRealmEntity is created and only in case `id` is not `null`.